Fix false positive external links in sitemaps

maartenbrakkee opened the pull request:
## Summary

This PR can be summarized in the following changelog entry:

* Fix false positive external links in sitemaps. Links to custom post types with `with_front` set to `false` were marked as external.

## Test instructions

This PR can be tested by following these steps:

* To reproduce follow the steps in #9634
* This pull should fix that

## Quality assurance

* [x] I have tested this code to the best of my abilities
* [ ] I have added unittests to verify the code works as intended

Fixes #9634

stodorovic commented:
Hi @maartenbrakkee,

First off, thank you for this PR. There are a few things I'd like to you change in the code:

- You have changed class _WPSEO_Link_Type_Classifier_, but from your description you should change class _WPSEO_Post_Type_Sitemap_Provider_ (method _get_classifier_). It's reason why phpunit tests are failed (PHPCS is failed too because you have used spaces instead tabs).

- I think that we should use function [_get_post_type_object_](https://codex.wordpress.org/Function_Reference/get_post_type_object) to retrieve _with_front_ flag for particular _post_type_ and remove path only if it is needed.

- You should added phpunit tests to confirm that this code works correctly.

I was thinking for possible solution related to #11714, #9634 and I think that we should create new method which returns _base_url_ for link classifier (but it could be an issue related to WPML). Another option is changing _get_home_url_ (in this case we should replace _get_home_url_ with _home_url_ in method _get_firsk_links_).

Thanks in advance.

maartenbrakkee commented:
Hi @stodorovic,

Thank you for your response. I've made some changes and updated the pull request. 

- I reverted my changes to _WPSEO_Link_Type_Classifier_. There are no changes to this class now.
- I made changes to _WPSEO_Post_Type_Sitemap_Provider_. It now has an extra private variable (boolean) which stores if the with_front rewrite setting, for the post type, was set to false or not. This way it should only be checked once for each post type that is being looped through. The boolean is set with the help of the result of the [get_post_type_object](https://codex.wordpress.org/Function_Reference/get_post_type_object) function.
- IDE extended to use .editorconfig. Tabs are now tabs and not spaces.

About the phpunit tests, I receive some weird errors when running `composer install`, which I can't seem to solve. Maybe you know the source of this?

```
> @php ./vendor/humbug/php-scoper/bin/php-scoper add-prefix --prefix=YoastSEO_Vendor --output-dir=./vendor_prefixed/ruckusing --config=config/php-scoper/ruckusing.inc.php --force --quiet

In Filesystem.php line 105:

  Failed to create "C:\web\wordpress-seo\./vendor_prefixed/ruckusingC:\web\wordpress-seo\vendor\ruckusing\ruckusing-migrations": mkdir(): No such file or directory.
```

maartenbrakkee commented:
Travis errors fixed.

stodorovic commented on inc/sitemaps/class-post-type-sitemap-provider.php:
```diff
@@ -45,13 +45,41 @@ class WPSEO_Post_Type_Sitemap_Provider implements WPSEO_Sitemap_Provider {
+	 * @return void
+	 */
+	protected function set_post_type_has_front( $post_type ) {
+		$post_type_has_front = get_post_type_object( $post_type )->rewrite['with_front'];
```
It's possible that `rewrite['with_front']` is not set. Also, we should check is it working in PHP 5.2.4. You should improve "code readability" ( I think that Yoast codebase doesn't use function dereferencing).

maartenbrakkee replied:
Updated in latest commit.

stodorovic commented on inc/sitemaps/class-post-type-sitemap-provider.php:
```diff
@@ -626,7 +656,9 @@ protected function get_url( $post ) {
-		if ( $this->get_classifier()->classify( $url['loc'] ) === WPSEO_Link::TYPE_EXTERNAL ) {
+		$post_type_has_front = $this->get_post_type_has_front();
+
+		if ( $this->get_classifier()->classify( $url['loc'] ) === WPSEO_Link::TYPE_EXTERNAL && $post_type_has_front ) {
```
I'd like to we don't touch this code. It's possible to pass _home_url_ without path (if with_front is true) in  instantiation  of_ _WPSEO_Link_Type_Classifier_ (it runs only once).

maartenbrakkee replied:
The problem is that _home_url_ and _site_url_ both contain the path, see [#9634](https://github.com/Yoast/wordpress-seo/issues/9634). For with_front set to true this is fine, but with false it is not.

I've updated it to remove the path from _home_url_ if the post type with_front rewrite option is set to false.

maartenbrakkee commented:
Travis errors fixed and updated based on feedback.

stodorovic commented:
@maartenbrakkee I've tried to test this PR and I've tried to create php unit tests (which should confirm that your PR works), but I wasn't able to reproduce your original issue. Could you update #9634 with more details about web server/proxy?
